lcls_tools/common/data_analysis/bmad_modeling.py

```diff
@@ -0,0 +1,573 @@
+                pvdata_reg[f"{pv}:P{attr}"] = phasF[indx]
+                pvdata_reg[f"{pv}:A{attr}"] = amplF[indx]
+            data_map = {"cavities": self.all_data_maps["cavities"]}
+            # datamap['quad'] =  self.all_data_maps['quad']}
+
+            tao_cmds = self.get_tao_cmds(pvdata_reg, data_map)
+            self.evaluate_tao(tao, tao_cmds)
+            region_e_tot = tao.ele_gen_attribs(f"END{region}")["E_TOT"] / 1e9
+            print(region_e_tot)
+
+        if "cu" in self.beam_path:
+            if region == "L2":
+                l2_edes = pvdata["BEND:LI24:790:EDES"] * 1e9
+                tao.cmd(f"set dat BC1.energy[2]|meas ={l2_edes} ")
+                tao.cmds(
+                    [
+                        "veto dat *",
+                        "veto var *",
+                        "use dat BC2.energy[1]",
+                        "use var linac_fudge[2]",
+                    ]
+                )
+            if region == "L3":
+                if "hxr" in self.beam_path:
+                    l3_edes = pvdata["BEND:DMPH:400:EDES"] * 1e9
+                elif "sxr" in self.beam_path:
+                    l3_edes = pvdata["BEND:DMPS:400:EDES"] * 1e9
+                tao.cmd(f"set dat L3[2]|meas ={l3_edes} ")
+                tao.cmds(
+                    [
+                        "veto dat *",
+                        "veto var *",
+                        "use dat L3.energy[2]",
+                        "use var linac_fudge[3]",
+                    ]
+                )
+            r = tao.cmd("run")
+            print(r)
+            for line in tao.cmd("show merit"):
+                print(line)
+
+    def get_tao_cmds(self, pvdata, data_maps):
+        """Returns command lines to update the Bmad model with Tao
+        where element values come from pvdata"""
+        lines = []
+        for dm_key in data_maps.keys():
+            lines += data_maps[dm_key].as_tao(pvdata)
+        return lines
+
+    def get_live(self, pvlist):
+        """Returns dictionary with PV names as keys and values of PVs"""
+        return dict(zip(pvlist, epics.caget_many(pvlist)))
+
+    def evaluate_tao(self, tao, tao_cmds):
+        """Evaluate set of tao command, toggles lattice_calculation OFF/ON
+        in between command list"""
+        tao.cmd("set global lattice_calc_on = F")
+        tao.cmds(tao_cmds)
+        tao.cmd("set global lattice_calc_on = T")
+        output = self.get_output(tao)
+        return output
+
+    def kmod_to_bdes(self, tao, element, K=0):
+        """Returns BDES given K1 in the Bmad model"""
+        ele = tao.ele_gen_attribs(element)
+        bp = ele["E_TOT"] / 1e9 / 299.792458 * 1e4  # kG m
+        return ele["K1"] * bp * ele["L"]
+
+    def bdes_to_kmod(self, tao, element, bdes):
+        """Returns K1 given BDES"""
+        ele = tao.ele_gen_attribs(element)
+        bp = ele["E_TOT"] / 1e9 / 299.792458 * 1e4
+        # kG m
+        return bdes / ele["L"] / bp  # kG / m / kG m = 1/m^2
+
```

Comment:
We are trying to keep plotting separate from basic data pulling or device code. Not sure if that needs to hold here too.

Reply:
I think here you mean this should be a separate function on a separate file?